File: 2095-minimum-number-of-swaps-to-make-the-string-balanced/minimum-number-of-swaps-to-make-the-string-balanced.java
class Solution {

    public int minSwaps(String s) {
    int swaps = 0;
    int balance = 0; // This will keep track of the balance between '[' and ']'
    
    for (char ch : s.toCharArray()) {
        if (ch == '[') {
            balance++;
        } else {
            balance--;
        }

        // If balance goes negative, it means we have more ']' than '[' at this point
        if (balance < 0) {
            swaps++;
            balance = 1; // Reset balance to 1 because we are assuming we did a swap
        }
    }
    
    return swaps;
}

            

}

        
    

File: 2095-minimum-number-of-swaps-to-make-the-string-balanced/minimum-number-of-swaps-to-make-the-string-balanced.py
class Solution:
    def minSwaps(self, s: str) -> int:
        balance = 0
        min_balance = 0
        for char in s:
            if char == '[':
                balance += 1
            else:
                balance -= 1
            min_balance = min(min_balance, balance)

        ans = -min_balance
        return (ans + 1) // 2
        